File: logcatviewer/src/main/java/com/github/logviewer/ReadLogcat.java
package com.github.logviewer;

import android.util.Log;

import androidx.annotation.NonNull;

import com.github.logviewer.databinding.LogcatViewerActivityLogcatBinding;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Scanner;
import java.util.regex.Pattern;

class ReadLogcat {
    private final List<Pattern> mExcludeList = new ArrayList<>();
    private ArrayList<String> excludeList;
    private final LogcatAdapter mAdapter = new LogcatAdapter();
    private boolean mReading = false;
    private Date time = null;

    public LogcatAdapter getAdapter() {
        return mAdapter;
    }

    public ArrayList<String> getExcludeList() {
        return excludeList;
    }

    public void setExcludeList(@NonNull ArrayList<String> list) {
        mExcludeList.clear();
        excludeList = list;
        for (String pattern : list) {
            try {
                mExcludeList.add(Pattern.compile(pattern));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void startReadLogcat(LogcatViewerActivityLogcatBinding mBinding) {
        new Thread() {
            @Override
            public void run() {
                super.run();
                mReading = true;
                try {
                    ArrayList<String> cmd = new ArrayList<>(Arrays.asList("logcat", "-v", "threadtime"));
                    if (time != null) {
                        SimpleDateFormat sdf = new SimpleDateFormat("MM-dd HH:mm:ss.mmm", Locale.getDefault());
                        cmd.add("-T");
                        cmd.add(sdf.format(time));
                    }
                    Log.d("logcat", "run: " + cmd);
                    Process process = new ProcessBuilder(cmd).start();
                    Scanner reader = new Scanner(process.getInputStream());
                    while (mReading && reader.hasNext()) {
                        String line = reader.nextLine();
                        if (LogItem.IGNORED_LOG.contains(line) || skip(mExcludeList, line)) {
                            continue;
                        }
                        try {
                            final LogItem item = new LogItem(line);
                            time = item.time;
                            mBinding.list.post(() -> mAdapter.append(item));
                        } catch (ParseException | NumberFormatException | IllegalStateException e) {
                            e.printStackTrace();
                        }
                    }
                    process.destroy();
                    reader.close();
                    mReading = false;
                    Log.d("logcat", "read logcat exit");
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }.start();
    }

    private boolean skip(@NonNull List<Pattern> mExcludeList, String line) {
        for (Pattern pattern : mExcludeList) {
            if (pattern.matcher(line).matches()) {
                return true;
            }
        }
        return false;
    }

    public boolean running() {
        return mReading;
    }

    public void stopReadLogcat() {
        mReading = false;
    }
}

File: logcatviewer/src/main/kotlin/com/github/logviewer/ReadLogcat.kt
package com.github.logviewer

import android.content.Context
import android.content.Intent
import android.view.MenuItem
import androidx.appcompat.widget.Toolbar
import androidx.core.content.FileProvider
import com.github.logviewer.databinding.LogcatViewerActivityLogcatBinding
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern

class ReadLogcat(
    private val context: Context,
    private val mBinding: LogcatViewerActivityLogcatBinding,
    excludeList: ArrayList<String>,
    private val floatingWindowLauncher: () -> Unit = emptyFunction(),
) : Toolbar.OnMenuItemClickListener {
    private val mExcludeList: MutableList<Pattern> = ArrayList()
    private var adapter: LogcatAdapter =
        LogcatAdapter(floatingWindowLauncher == emptyFunction(), mBinding)
    private var mReading = false
    private var time: Date? = null

    init {
        mBinding.list.adapter = this@ReadLogcat.adapter
        mExcludeList.clear()
        excludeList.forEach {
            ignore {
                mExcludeList.add(Pattern.compile(it))
            }
        }
    }

    @DelicateCoroutinesApi
    fun start() {
        GlobalScope.launch(Dispatchers.IO) {
            mReading = true
            val cmd = ArrayList(listOf("logcat", "-v", "threadtime"))
            time?.let {
                cmd.add("-T")
                cmd.add(SimpleDateFormat(DATE_FORMAT, Locale.getDefault()).format(it))
            }

            val process = ProcessBuilder(cmd).start()

            process.inputStream.bufferedReader().use {
                while (mReading)
                    it.readLine()?.let { line ->
                        if (LogItem.IGNORED_LOG.contains(line) || skip(mExcludeList, line))
                            return@let
                        ignore {
                            LogItem(line).let { item ->
                                time = item.time
                                adapter.append(item)
                            }
                        }
                    } ?: break
            }

            process.destroy()
            stop()
        }
    }

    private fun skip(mExcludeList: List<Pattern>, line: String): Boolean {
        for (pattern in mExcludeList) if (pattern.matcher(line).matches()) return true
        return false
    }

    fun running(): Boolean = mReading

    fun stop() {
        mReading = false
    }

    @OptIn(DelicateCoroutinesApi::class)
    override fun onMenuItemClick(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.clear -> adapter.clear()
            R.id.export -> exportLogFile()
            R.id.floating -> floatingWindowLauncher()
            R.id.Verbose, R.id.Debug, R.id.Info, R.id.Warning, R.id.Error, R.id.Fatal ->
                adapter.setFilter(item.title?.get(0).toString())
            else -> return false
        }
        return true
    }

    @DelicateCoroutinesApi
    fun exportLogFile() {
        GlobalScope.launch(Dispatchers.IO) {
            File(context.externalCacheDir, "yuhaiin.log").apply {
                writeText("Yuhaiin Logcat:\n")
                Runtime.getRuntime()
                    .exec(arrayOf("logcat", "-d")).inputStream.use { input ->
                        FileOutputStream(this, true).use {
                            input.copyTo(it)
                        }
                    }

                val authority = "${context.packageName}.logcat_fileprovider"
                val uri = FileProvider.getUriForFile(context, authority, this)
                context.startActivity(
                    Intent.createChooser(
                        Intent(Intent.ACTION_SEND)
                            .setType("text/plain")
                            .setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                            .putExtra(Intent.EXTRA_STREAM, uri),
                        "Export Logcat"
                    ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                )
            }
        }
    }

    companion object {
        const val CONTENT_TEMPLATE =
            "Time: %s\nPid: %d Tid: %d Priority: %s Tag: %s\n\nContent: \n%s"
        const val DATE_FORMAT = "MM-dd HH:mm:ss.SSS"

        fun ignore(f: () -> Unit) {
            try {
                f()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        private fun emptyFunction(): () -> Unit = {}
    }
}